File: src/com/medium/KthLargestElementinanArray_215.java
/**
 * 
 */
package com.medium;

import java.util.Arrays;
import java.util.PriorityQueue;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class KthLargestElementinanArray_215 {

    public static void main(String[] args) {
        int[] nums = { 3, 2, 3, 1, 2, 4, 5, 5, 6 };// {3,2,1,5,6,4};
        int k = 4;// 2;

        System.out.println(findKthLargest(nums, k));
    }

    public static int findKthLargest(int[] nums, int k) {

        // return priorityQueue(nums, k);
        // return libSort(nums, k);
        return sort(nums, k);
    }

    // Use priority queue
    // Time Complexity: O(nlogk)
    // Space Complexity:O(k)
    // Runtime: 6 ms, faster than 64.98%
    // Memory Usage: 39.7 MB, less than 5.17%
    private static int priorityQueue(int[] nums, int k) {
        PriorityQueue<Integer> mPriorityQueue = new PriorityQueue<>();
        for (int i = 0; i < nums.length; i++) {
            mPriorityQueue.add(nums[i]);
            if (mPriorityQueue.size() > k) {
                mPriorityQueue.remove(); // also can use poll()
            }
        }
        return mPriorityQueue.element(); // also can use peek()

    }

    // Use build-in sort lib
    // Time Complexity: O(nlogn)
    // Space Complexity:O(1)
    // Runtime: 2 ms, faster than 97.64%
    // Memory Usage: 38 MB, less than 92.53%
    private static int libSort(int[] nums, int k) {
        Arrays.sort(nums); // use lib sort
        return nums[nums.length - k];
    }

    private static int sort(int[] nums, int k) {
        quickSort(nums, 0, nums.length - 1);
        // for(int i = 0; i < nums.length;i++) {
        // System.out.printf("%d ",nums[i]);
        // }
        return nums[nums.length - k];
    }

    // Use Quick sort with DS version
    // Time Complexity: O(nlogn)
    // Space Complexity:O(1)
    // Runtime: 28 ms, faster than 35.46%
    // Memory Usage: 39.2 MB, less than 13.73%
    private static void quickSort(int[] nums, int mIndexStart, int mIndexEnd) {
        if (mIndexStart < mIndexEnd) {

            // System.out.println("test");
            int mPivot, mTmpIndexLeft = mIndexStart, mTmpIndexRight = mIndexEnd;
            mPivot = nums[mIndexStart];
            while (mTmpIndexLeft < mTmpIndexRight) {

                while ((nums[mTmpIndexLeft] <= mPivot) && (mTmpIndexLeft < mIndexEnd)) {
                    mTmpIndexLeft++;
                }
                while ((nums[mTmpIndexRight] > mPivot) && (mTmpIndexRight > mIndexStart)) {
                    mTmpIndexRight--;
                }

                if (mTmpIndexLeft < mTmpIndexRight) {
                    // System.out.printf("IndexLeft:%d
                    // IndexRight:%d\n",mTmpIndexLeft, mTmpIndexRight);
                    swap(nums, mTmpIndexLeft, mTmpIndexRight);
                }
            }

            swap(nums, mIndexStart, mTmpIndexRight);

            // System.out.printf("mIndexStart:%d mIndexEnd:%d\n", mIndexStart,
            // mIndexEnd);
            // System.out.printf("mTmpIndexLeft:%d mTmpIndexRight:%d\n",
            // mTmpIndexLeft, mTmpIndexRight);

            quickSort(nums, mIndexStart, (mTmpIndexRight - 1));
            quickSort(nums, (mTmpIndexRight + 1), mIndexEnd);
        }
    }

    private static int partition(int[] nums, int mIndexLeft, int mIndexRight) {
        int i = mIndexLeft - 1;
        for (int j = mIndexLeft; j < mIndexRight; j++) {
            if (nums[j] < nums[mIndexRight]) {
                i++;
                swap(nums, i, j);
            }
        }
        swap(nums, (i + 1), mIndexRight);
        return (i + 1);
    }

    // Quick sort with Algorithm version (partition)
    // Time Complexity: O(nlogn)
    // Space Complexity:O(1)
    // Runtime: 84 ms, faster than 5.22%
    // Memory Usage: 38.8 MB, less than 57.34%
    private static void quickSort2(int[] nums, int mIndexLeft, int mIndexRight) {
        if (mIndexLeft < mIndexRight) {
            int q = partition(nums, mIndexLeft, mIndexRight);
            quickSort2(nums, mIndexLeft, q - 1);
            quickSort2(nums, q + 1, mIndexRight);
        }
    }

    private static void swap(int[] nums, int mIndex1, int mIndex2) {
        // System.out.printf("%d %d\n", mIndex1, mIndex2);
        int mTmp = nums[mIndex1];
        nums[mIndex1] = nums[mIndex2];
        nums[mIndex2] = mTmp;
    }

    // Use quick select (partition concept)
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 16 ms, faster than 36.87%
    // Memory Usage: 36.7 MB, less than 90.67%
    public int findKthLargest2(int[] nums, int k) {
        return quickSelect(nums, 0, nums.length - 1, k);
    }

    private int quickSelect(int[] nums, int mIndexLow, int mIndexHigh, int k) {
        int i = mIndexLow - 1;

        for (int j = mIndexLow; j < mIndexHigh; j++) {
            if (nums[j] < nums[mIndexHigh]) {
                i++;
                swap(nums, i, j);
            }
        }
        swap(nums, i + 1, mIndexHigh);

        int mCount = mIndexHigh - (i + 1) + 1;
        if (mCount == k) {
            return nums[i + 1];
        } else if (mCount > k) { // pivot is small, so it must be on the right
            return quickSelect(nums, (i + 1) + 1, mIndexHigh, k);
        } else { // pivot is big, so it must be on the left
            return quickSelect(nums, mIndexLow, (i + 1) - 1, k - mCount);
        }
    }
}

File: Golang_src/medium/kthLargestElementInAnArray_215.go
package medium

// Use quick select:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 32 ms, faster than 6.04%
// Memory Usage: 4 MB, less than 35.96%
func findKthLargest(nums []int, k int) int {
	return quickSelect(&nums, 0, len(nums)-1, k)
}

func quickSelect(nums *[]int, indexLow, indexHigh, k int) int {
	i := indexLow - 1

	for j := indexLow; j < indexHigh; j++ {
		if (*nums)[j] < (*nums)[indexHigh] {
			i++
			(*nums)[i], (*nums)[j] = (*nums)[j], (*nums)[i]
		}
	}
	i++
	(*nums)[i], (*nums)[indexHigh] = (*nums)[indexHigh], (*nums)[i]

	count := indexHigh - i + 1
	if count == k {
		return (*nums)[i]
	} else if count > k {
		return quickSelect(nums, i+1, indexHigh, k)
	} else {
		return quickSelect(nums, indexLow, i-1, k-count)
	}
}

// Use quick select 2:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 32 ms, faster than 5.33%
// Memory Usage: 3.9 MB, less than 42.13%
func findKthLargest2(nums []int, k int) int {
	var qSelect func(start, end, k int) int
	qSelect = func(start, end, k int) int {
		i := start - 1

		// O(n)
		for j := start; j < end; j++ {
			if nums[j] < nums[end] {
				i++
				nums[i], nums[j] = nums[j], nums[i]
			}
		}
		i++
		nums[end], nums[i] = nums[i], nums[end]

		cnt := end - i + 1
		if cnt == k {
			return nums[i]
		} else if cnt < k {
			return qSelect(start, i-1, k-cnt)
		} else {
			return qSelect(i+1, end, k)
		}
	}

	return qSelect(0, len(nums)-1, k)
}

// Use quick select with descending order:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 81 ms, faster than 70.8%
// Memory Usage: 8.6 MB, less than 76.64%
func findKthLargest3(nums []int, k int) int {
	var qs func(l, r, k int) int
	qs = func(l, r, k int) int {
		i := l - 1
		for j := l; j < r; j++ {
			if nums[j] > nums[r] {
				i++
				nums[i], nums[j] = nums[j], nums[i]
			}
		}
		i++
		nums[r], nums[i] = nums[i], nums[r]

		cnt := i + 1

		if cnt == k {
			return nums[i]
		} else if cnt < k {
			return qs(i+1, r, k)
		} else {
			return qs(l, i-1, k)
		}
	}

	return qs(0, len(nums)-1, k)
}
